src/librustc_mir/transform/util.rs

```diff
@@ -0,0 +1,46 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use repr::*;
+
+/// Update basic block ids in all terminators using the given replacements,
+/// useful e.g. after removal of several basic blocks to update all terminators
+/// in a single pass
+pub fn update_basic_block_ids(mir: &mut Mir, replacements: &[usize]) {
+    for bb in mir.all_basic_blocks() {
+        for target in mir.basic_block_data_mut(bb).terminator.successors_mut() {
+            *target = BasicBlock::new(replacements[target.index()]);
+        }
+    }
+}
+
+/// Mass removal of basic blocks to keep the ID-remapping cheap.
+pub fn remove_basic_blocks(mir: &mut Mir, keep: &[bool]) {
+    // Check that we have a usage flag for every block
+    assert_eq!(mir.basic_blocks.len(), keep.len());
+
+    // `mapping` keeps track which new block id refers to which old block id
+    let mut mapping: Vec<_> = (0..mir.basic_blocks.len()).collect();
+    // `replacements` maps the old block ids to the new ones
+    let mut replacements: Vec<_> = mapping.clone();
+
+    for (i, keep) in keep.iter().enumerate().rev() {
+        if !keep {
+            mapping.swap_remove(i);
+            mir.basic_blocks.swap_remove(i);
```

Comment:
So another way to do this would be (pseudocode):

    let N = mir.basic_blocks.len();
    let mut replacements = (0..N).collect();
    let mut dead = 0;
    for i in 0..N {
        if keep[i] && dead > 0 {
            replacements[i] -= dead;
            mir.basic_blocks.swap(i, i - dead);
        } else {
            dead += 1;
        }
    }
    mir.basic_blocks.truncate(N - dead);

I think this works. The basic idea is to compact the live blocks into the front. The reason this might be better is that it preserves the relative ordering of the blocks. Not that this really matters per se.